content/courses/intro-to-embedded-sys-prog/chapters/interacting_with_devices.rst

```diff
@@ -3,152 +3,1633 @@ Interacting with Devices
+
+Why do we need 4-bit values? At the hardware level, each memory-mapped
+GPIO port has a sequence of 16 4-bit quantities, one for each of the 16
+pins on the port. Those 4-bit quantities specify the "alternate
+functions" that the pin can take on, if needed. The alternate functions
+allow a given pin to do more than act as a single discrete I/O pin. For
+example, a pin could be connected to the incoming lines of a USART. We
+use the configuration routine to apply the specific code representing
+the alternate function required for our application.
+
+These 16 4-bit alternate function code fields are contiguous in the
+register (hence memory) so we can represent them as an array with a
+total size of 64-bits (i.e., 16 times 4). In the C version this array
+has as two components of type :c:`uint32_t` so it must compute where the
+corresponding 4-bit value for the pin is located within those two words.
+In contrast, the Ada version of the array has components of the 4-bit
+type, rather than two 32-bit components, and simply uses the pin number
+as the index. The resulting Ada procedure body is extremely simple:
+
+.. code-block:: ada
+
+   procedure Configure_Alternate_Function
+     (Port : in out GPIO_Port;
+      Pin  : GPIO_Pin;
+      AF   : GPIO_Alternate_Function_Code)
+   is
+   begin
+      Port.AFR (Pin) := AF;
+   end Configure_Alternate_Function;
+
+In the Ada version, :ada:`AFR` is a component within the
+:ada:`GPIO_Port` record type, much like in the C code's struct. However,
+Ada allows us to declare a much more descriptive set of types, and it is
+these types that allows the developer to off-load the work to the compiler.
+
+First, in Ada we can declare a 4-bit numeric type:
+
+.. code-block:: ada
+
+   type Bits_4 is mod 2**4 with Size => 4;
+
+The :ada:`Bits_4` type was already globally defined elsewhere so we just
+derive our 4-bit "alternate function code" type from it. Doing so allows the
+compiler to enforce simple strong typing so that the two value spaces
+are not accidentally mixed. This approach also increases understanding
+for the reader:
+
+.. code-block:: ada
+
+   type GPIO_Alternate_Function_Code is new Bits_4;
+   --  We cannot use an enumeration type because there are duplicate binary
+   --  values
+
+Hence type :ada:`GPIO_Alternate_Function_Code` is a copy of
+:ada:`Bits_4` in terms of operations and values, but is not the same
+type as :ada:`Bits_4` so the compiler will keep them separate for us.
+
+We can then use that type as the array component type for the representation
+of the :ada:`AFR`:
+
+.. code-block:: ada
+
+   type Alternate_Function_Fields is array (GPIO_Pin) of GPIO_Alternate_Function_Code
+     with Component_Size => 4, Size => 64;  -- both in units of bits
+
+Note that we can use the GPIO :ada:`Pin` parameter directly as the index into
+the array type, obviating any need to massage the :ada:`Pin` value in
+the procedure. That's possible because the type :ada:`GPIO_Pin` is an
```

Comment:
``GPIO_Pin`` isn't defined in any of the code examples. Maybe we should have a type declaration here?

Reply:
Will do.